typed-protocols/src/Network/TypedProtocol/Driver/ByteLimit.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Network.TypedProtocol.Driver.ByteLimit
+  ( runPeerWithByteLimit
+  , DecoderFailureOrTooMuchInput (..)
+  ) where
+
+import           Control.Exception
+import           Data.Int (Int64)
+
+import           Control.Monad.Class.MonadThrow (MonadThrow (..))
+import           Control.Tracer (Tracer, traceWith)
+
+
+import           Network.TypedProtocol.Core
+import           Network.TypedProtocol.Codec
+import           Network.TypedProtocol.Channel
+import           Network.TypedProtocol.Driver
+
+
+data DecoderFailureOrTooMuchInput failure
+  = DecoderFailure !failure
+  | TooMuchInput
+  deriving (Show)
+
+instance Exception failure => Exception (DecoderFailureOrTooMuchInput failure)
+
+runDecoderWithByteLimit
+    :: forall m bytes failure a. Monad m
+    => Int64
+    -- ^ message size limit
+    -> (bytes -> Int64)
+    -- ^ byte size
+    -> Channel m bytes
+    -> Maybe bytes
+    -> DecodeStep bytes failure m a
+    -> m (Either (DecoderFailureOrTooMuchInput failure) (a, Maybe bytes))
+runDecoderWithByteLimit limit byteLength Channel{recv} mbytes = go 0 mbytes
+    where
+      go :: Int64
+         -- ^ length of consumed input
+         -> Maybe bytes
+         -> DecodeStep bytes failure m a
+         -> m (Either (DecoderFailureOrTooMuchInput failure) (a, Maybe bytes))
+      -- we decoded the data, but we might be over the limit
+      go !l _  (DecodeDone x trailing) | l - maybe 0 byteLength trailing > limit = return (Left TooMuchInput)
+                                       | otherwise                       = return (Right (x, trailing))
+      -- we run over the limit, return @TooMuchInput@ error
+      go !l _  _                       | l > limit = return (Left TooMuchInput)
+      go !_ _  (DecodeFail failure)    = return (Left (DecoderFailure failure))
+
+      go !l Nothing         (DecodePartial k) =
+        recv >>= (\mbs -> k mbs >>= go (l + maybe 0 byteLength mbs) Nothing)
+
+      go !l (Just trailing) (DecodePartial k) =
+        k (Just trailing) >>= go (l + byteLength trailing) Nothing
+
+
+-- |
+-- Like @'runPeer'@, but require that each inbound message is smaller than
+-- @limit@ bytes; if the limit is exceeded throw @'TooMuchInput'@ exception.
+--
+runPeerWithByteLimit
+  :: forall ps (st :: ps) pr bytes failure m a .
+     (MonadThrow m, Exception failure)
+  => Int64
+  -- ^ message size limit
+  -> (bytes -> Int64)
+  -- ^ byte size
+  -> Tracer m (TraceSendRecv ps)
+  -> Codec ps failure m bytes
+  -> Channel m bytes
+  -> Peer ps pr st m a
+  -> m a
+
+runPeerWithByteLimit limit byteLength tr Codec{encode, decode} channel@Channel{send} =
```

Comment:
Ideally we'd be able to use the same original `runPeer`, but with a _codec_ that enforces the byte limit. Is this not feasible for some reason? I'd guess that maybe cborg doesn't offer this feature, of counting the bytes consumed so far?